eng/targets/GeneratePkgDef.targets

```diff
@@ -0,0 +1,221 @@
+         
+         Project may specify <PkgDefEntry>BindingRedirect</PkgDefEntry> on any item that contributes to ReferencePath item group.
+         (e.g. ProjectReference), or on NuGetPackageToIncludeInVsix items. These items will be automatically included in PkgDefEntry items.
+         
+      4) CodeBase
+         Emits $RootKey$\RuntimeConfiguration\dependentAssembly\codeBase entry.
+         
+         Project may specify <PkgDefEntry>CodeBase</PkgDefEntry> on any item that contributes to ReferencePath item group.
+         (e.g. ProjectReference), or on NuGetPackageToIncludeInVsix items. These items will be automatically included in PkgDefEntry items.
+         
+      5) FileContent
+         Merges the content of the file whose path is specified in the identity of the item to the generated pkgdef file.
+         The path may be relative to the project being built.
+    
+    The FileContent kind allows the project to add arbitrary static content. The other kinds are used to generate dynamic content.
+  -->
+  
+  <UsingTask TaskName="Microsoft.DotNet.Build.Tasks.VisualStudio.GetPkgDefAssemblyDependencyGuid" AssemblyFile="$(ArcadeVisualStudioBuildTasksAssembly)" />
+  <UsingTask TaskName="Microsoft.DotNet.Arcade.Sdk.GetAssemblyFullName" AssemblyFile="$(ArcadeSdkBuildTasksAssembly)" />
+    
+  <Target Name="_SetGeneratePkgDefInputsOutputs">
+    <PropertyGroup>
+      <!-- The path must match the value that VSSDK uses. -->
+      <_GeneratePkgDefOutputFile>$(IntermediateOutputPath)$(TargetName).pkgdef</_GeneratePkgDefOutputFile>
+    </PropertyGroup>
+
+    <ItemGroup>
+      <_ContentEntries Include="@(Content)" Condition="'%(Content.PkgDefEntry)' != ''">
+        <FullFilePath>$([System.IO.Path]::Combine('$(MSBuildProjectDirectory)', '%(Content.Identity)'))</FullFilePath>
+      </_ContentEntries>
+
+      <PkgDefEntry Include="@(_ContentEntries->'%(FullFilePath)')" >
+        <Kind>%(_ContentEntries.PkgDefEntry)</Kind>
+      </PkgDefEntry>
+    </ItemGroup>
+  </Target>
+  
+  <!--
+    Initializes metadata on PkgDefEntry items with Kind="PackageRegistration".
+  -->
+  <Target Name="_InitializePackageRegistrationEntries"
+          BeforeTargets="GeneratePkgDef"
+          DependsOnTargets="GetAssemblyVersion">
+    <PropertyGroup>
+      <_AssemblyCodeBase>$([System.IO.Path]::Combine('$PackageFolder$', '$(AssemblyVSIXSubPath)', '$(TargetFileName)'))</_AssemblyCodeBase>
+      <_AssemblyOrCodeBase Condition="'$(UseCodeBase)' != 'true'">"Assembly"="$(AssemblyName), Version=$(AssemblyVersion), Culture=neutral, PublicKeyToken=$(PublicKeyToken)"</_AssemblyOrCodeBase>
+      <_AssemblyOrCodeBase Condition="'$(UseCodeBase)' == 'true'">"CodeBase"="$(_AssemblyCodeBase)"</_AssemblyOrCodeBase>
+    </PropertyGroup>
+    <ItemGroup>
+      <PkgDefEntry Update="@(PkgDefEntry)" Condition="'%(PkgDefEntry.Kind)' == 'PackageRegistration' and '%(PkgDefEntry.AssemblyName)' == ''">
```

Comment:
Was thinking about that... probably would be better at this point. Initially I only hard binding redirects and code bases, which share a lot of common logic. But now that we have 5 kinds I think it would be simpler to do as you suggest.